lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp

```diff
@@ -0,0 +1,261 @@
+                                              ArrayRef<Value> operands,
+                                              ArrayRef<Type> resultTypes) {
+  OperationState res(loc, op->getName().getStringRef(), operands, resultTypes,
+                     op->getAttrs());
+  return builder.createOperation(res);
+}
+
+/// Helper to create a new WarpSingleLaneOp regions with different signature.
+static WarpSingleLaneOp
+moveRegionToNewWarpSignature(OpBuilder &b, WarpSingleLaneOp warpOp,
+                             ValueRange newYieldedValues,
+                             TypeRange newReturnTypes) {
+  // Create a new op before the existing one, with the extra operands.
+  OpBuilder::InsertionGuard g(b);
+  b.setInsertionPoint(warpOp);
+  auto newWarpOp = b.create<WarpSingleLaneOp>(warpOp.getLoc(), newReturnTypes,
+                                              warpOp.laneid());
+
+  Region &opBody = warpOp.getBodyRegion();
+  Region &newOpBody = newWarpOp.getBodyRegion();
+  newOpBody.takeBody(opBody);
+  auto yield =
+      cast<vector_ext::YieldOp>(newOpBody.getBlocks().begin()->getTerminator());
+  yield.operandsMutable().assign(newYieldedValues);
+  return newWarpOp;
+}
+
+/// Helper to create a new WarpSingleLaneOp region with extra outputs.
+static WarpSingleLaneOp
+moveRegionToNewWarpAddReturn(OpBuilder &b, WarpSingleLaneOp warpOp,
+                             ValueRange newYieldedValues,
+                             TypeRange newReturnTypes) {
+
+  SmallVector<Type> types(warpOp.getResultTypes().begin(),
+                          warpOp.getResultTypes().end());
+  types.append(newReturnTypes.begin(), newReturnTypes.end());
+  auto yield = cast<vector_ext::YieldOp>(
+      warpOp.getBodyRegion().getBlocks().begin()->getTerminator());
+  SmallVector<Value> yieldValue(yield.getOperands().begin(),
+                                yield.getOperands().end());
+  yieldValue.append(newYieldedValues.begin(), newYieldedValues.end());
+  WarpSingleLaneOp newWarpOp =
+      moveRegionToNewWarpSignature(b, warpOp, yieldValue, types);
+  for (auto it :
+       llvm::zip(warpOp.getResults(),
+                 newWarpOp.getResults().take_front(warpOp.getNumResults())))
+    std::get<0>(it).replaceAllUsesWith(std::get<1>(it));
+  return newWarpOp;
+}
+
+llvm::Optional<std::pair<Operation *, unsigned>>
+getWarpResult(WarpSingleLaneOp warpOp, std::function<bool(Operation *)> fn) {
+  auto yield = cast<vector_ext::YieldOp>(
+      warpOp.getBodyRegion().getBlocks().begin()->getTerminator());
+  for (OpOperand &yieldOperand : yield->getOpOperands()) {
+    Value yieldValue = yieldOperand.get();
+    Operation *definedOp = yieldValue.getDefiningOp();
+    if (definedOp && fn(definedOp)) {
+      if (!warpOp.getResult(yieldOperand.getOperandNumber()).use_empty())
+        return std::make_pair(definedOp, yieldOperand.getOperandNumber());
+    }
+  }
+  return {};
+}
+
+// Currently the distribution map is implicit based on the vector shape. In the
```

Comment:
Please give an example of the current implicit behavior.
Also, seems to be more naturally expressible as an API on vector shapes?
```
static AffineMap calculateImplicitMap(ArrayRef<int64_t> shape, ArrayRef<int64_t> subShape) {
  int64_t i =0;
  for (auto it = llvm::zip(shape, subShape)) {
    assert(std::get<0>(it) % std::get<1>(it) == 0 ...);
    if (std::get<0>(it) != std::get<1>(it)) {
      AffineExpr d = getAffineDimExpr(i++, yield.getContext());
      perm.push_back(d * (std::get<0>(it) / std::get<1>(it))); // I think I'd put the scale in here too.
    }
  }
  return AffineMap::get(srcType.getRank(), 0, perm, yield.getContext());
}

Reply:
I added an example. I'm considering allowing non vector values in the future so keeping the value seems more flexible for now.